### PR TITLE
Implement QR solver in linalg library

### PR DESCRIPTION
**Pull request by SashaMalysheva**
I have implemented SVD and GR decomposition and solver.
The task is incomplete because I've encountered with some problems in replacing QR solver in LDA and FisherLDA.

ref #3716

### CODE REVIEW

**Comment by lisitsyn** [comment]
@SashaMalysheva what were the problems with LDA?

File: src/shogun/mathematics/linalg/LinalgBackendBase.h [hunk 1]
```diff
@@ -132,6 +132,93 @@ class LinalgBackendBase
 	DEFINE_FOR_NON_INTEGER_PTYPE(BACKEND_GENERIC_CHOLESKY_SOLVER, SGMatrix)
 	#undef BACKEND_GENERIC_CHOLESKY_SOLVER
 
+    /**
+     * Wrapper solver with QR decomposition.
+     *
+     * @see linalg::qr_solver
+     */
+    #define BACKEND_GENERIC_QR_SOLVER(Type, Container) \
+	virtual SGMatrix<Type> qr_solver(const Container<Type>& A, \
```
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendBase.h]: identation

Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendBase.h]: and elsewhere

**Comment by OXPHOS** [comment]
It seems like the `eigen.qr_solver` itself failed from the Travis. No?

**Comment by karlnapf** [comment]
Hi!
This looks quite good. However, you realise that if you want all matrices (U, S, V) of an SVD, it computes the SVD three times?
Instead, the method should return a triple with all matrices ...

**Pull request closed without merging**